### ISSUE TITLE
Attach type parameter to (non-method) function-valued property in object type

### ISSUE THREAD

**Issue by ethanresnick** [Microsoft/TypeScript#24401]
## Search Terms
generic in object key type alias parameter

## Suggestion

My apologies if this has been suggested before or if there's already a way to do it that I'm not aware of... I wasn't exactly sure how to search for this issue. But, basically, here's the situation:

Suppose I have an interface/object type like:

```ts
interface X {
  applyPreMiddleware<T extends SomeType = SomeType>(it: T, opts: B): Promise<T | undefined>
  applyPostMiddleware<T extends SomeTypePostProcessing = SomeTypePostProcessing>(it: T, opts: B): Promise<T | undefined>
}

// Note: SomeTypePostProcessing extends SomeType
```

The `applyPreMiddleware` and `applyPostMiddleware` accept an object and apply registered middleware functions for objects of that type. Now, suppose I want to pull out the common signature, say, to use for an underlying function. Something like: 

```ts
type applyMiddleware<T extends SomeType = SomeType> = 
  (it: T, opts: B) => Promise<T | undefined>
```

What I'd like is a way to use this common type to define the `interface X` above, so I don't have to repeat the common bits. Something like:

```ts
interface X {
  applyPreMiddleware<T extends SomeType = SomeType>: applyMiddleware<T>;
  applyPostMiddleware<T extends SomeTypePostProcessing = SomeTypePostProcessing>: applyMiddleware<T>;
}
```

Basically, I'm trying to reference the type alias instead of putting the function signature inline. When I try the above, though, I get a syntax error.

Note: the generic parameter has to be on the methods, not the interface, because the value for `T` can be different each time each method is called.

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by DanielRosenwasser** [comment]
I think you want this:

```ts
type ApplyMiddleware<Constraint> = 
  <T extends Constraint = Constraint>(it: T, opts: B) => Promise<T | undefined>

interface X {
  applyPreMiddleware: ApplyMiddleware<SomeType>;
  applyPostMiddleware: ApplyMiddleware<SomeTypePostProcessing>;
}
```

**Comment by ethanresnick** [comment]
```ts

interface X {
  applyPreMiddleware: ApplyMiddleware<SomeType>;
  applyPostMiddleware: ApplyMiddleware<SomeTypePostProcessing>;
}
```

Thanks @DanielRosenwasser.

If I'm understanding correctly, though, I don't think your suggestion works. In particular, I want to allow the user of the `X` to set the parameter at each call site of one of the methods; in your version, though, the value for the parameters seems to be set once and forever in the definition of `X`.

**Comment by ethanresnick** [comment]
@DanielRosenwasser Should this be relabeled/reopened? I'm pretty sure your suggestion doesn't work exactly for my use case.
